server/service/src/requisition/program_settings/map.rs

```diff
@@ -0,0 +1,93 @@
+use repository::{
+    PeriodRow, ProgramRequisitionOrderTypeRow, ProgramRequisitionSettings, RequisitionsInPeriod,
+};
+
+use super::{prepare::PrepareProgramSettings, OrderType, ProgramSettings};
+
+/// Map prepared program_settings, order_types, periods and requisitions_in_periods to ProgramSettings
+/// order types are mapped to program settings by program setting id
+/// available periods are mapped with period_is_available, see description
+/// supplier are mapped by program_id
+pub(super) fn map(
+    PrepareProgramSettings {
+        settings,
+        order_types,
+        periods,
+        program_suppliers,
+        requisitions_in_period,
+    }: PrepareProgramSettings,
+) -> Vec<ProgramSettings> {
+    settings
+        .into_iter()
+        .map(|program_setting| {
+            // Filter and map (to include available_periods) order type for program setting
+            let order_types = order_types
+                .iter()
+                .filter(|o| {
+                    o.program_requisition_settings_id == program_setting.program_settings_row.id
+                })
+                .map(|order_type| {
+                    // Filter periods for order type, period_is_available
+                    let available_periods = periods
+                        .iter()
+                        .filter(|period| {
+                            period_is_available(
+                                period,
+                                &program_setting,
+                                &order_type,
+                                &requisitions_in_period,
+                            )
+                        })
+                        .map(|p| p.clone())
```

Comment:
Hmm, i am kind of used to it not, and chaining (functional ?) seems to be a common practice now days.

Ohh yeah as for nested maps, I tend to agree, but this is kind of what I need to do, map a nested object structure. How would you do this, break it down to methods ? (`map_order_type` and `map_periods`) ?

Also I thought since it's only 40 lines, it was ok to do with map. Would suggest if this mapping gets any bigger, extract into methods

Reply:
Yeah I think something like `map_order_type` and `map_periods` would be good as it give semantic meaning your brain can take in more easily. We can then separately check if  how we map periods is making sense.
Like you say I think this is ok, but it's on the upper bounds of the complexity in my mind. Already hard to understand and as more complexity gets added could get harder to understand. Just thought I'd comment, but no real action needed.